### PR TITLE
Issue #11643: UnnecessaryParentheses: false negative on unnecessary p…

### PR DESCRIPTION
**Pull request by Kevin222004**
Issue #11643: UnnecessaryParentheses: false negative on unnecessary UnnecessaryParentheses

---------------------------------------------------------------------------
Diff Regression config: https://gist.githubusercontent.com/Kevin222004/e206f8b15b0d8c04e1d6b1fbded90dbe/raw/9e49e196b7a4503d162eeefba01bc4d0daa58ac3/config.xml

### CODE REVIEW

**Comment by Kevin222004** [comment]
Github, generate report

**Comment by Kevin222004** [comment]
@romani https://checkstyle.semaphoreci.com/jobs/0a3e69de-02ce-4515-86d9-470bc0be261f please look at this the same failure is at https://github.com/checkstyle/checkstyle/pull/12548 which is not related to changes

**Comment by Kevin222004** [comment]
@romani @nrmancuso I want to change https://github.com/checkstyle/checkstyle/blob/e1c2b4650b3ca419e88bdb6fcc95dac64a530aaa/.ci/validation.sh#L635
this with lastest commit but i guess i have to make changes in https://github.com/checkstyle/checkstyle/blob/e1c2b4650b3ca419e88bdb6fcc95dac64a530aaa/.ci/validation.sh#L624-L627
hipparchus as well i am not well aware of this can u please help here

**Comment by Kevin222004** [comment]
can you please look at https://gist.githubusercontent.com/Kevin222004/e206f8b15b0d8c04e1d6b1fbded90dbe/raw/9e49e196b7a4503d162eeefba01bc4d0daa58ac3/config.xml 
Diff report is not generating on local as well with this error
```
[INFO] ------------------------------------------------------------------------
Starting diff report generation ...
Running command: java -jar /home/kevin/Desktop/check_style/checkstyle/contribution/patch-diff-report-tool/target/patch-diff-report-tool-0.1-SNAPSHOT-jar-with-dependencies.jar --patchReport reports/UParanthesis/checkstyle/checkstyle-result.xml
                        --output reports/diff/checkstyle --patchConfig /home/kevin/Desktop/check_style/checkstyle/contribution/checkstyle-tester/my_check.xml --baseReport reports/master/checkstyle/checkstyle-result.xml --baseConfig /home/kevin/Desktop/check_style/checkstyle/contribution/checkstyle-tester/my_check.xml
patch-diff-report-tool execution started.
XML parsing is started.
Creation of configuration report is started.
Exception in thread "main" javax.xml.stream.XMLStreamException: ParseError at [row,col]:[1,2]
Message: The markup declarations contained or pointed to by the document type declaration must be well-formed.
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLStreamReaderImpl.next(XMLStreamReaderImpl.java:652)
        at java.xml/com.sun.xml.internal.stream.XMLEventReaderImpl.nextEvent(XMLEventReaderImpl.java:83)
        at com.github.checkstyle.parser.CheckstyleConfigurationsParser.parseModule(CheckstyleConfigurationsParser.java:147)
        at com.github.checkstyle.parser.CheckstyleConfigurationsParser.parseConfiguration(CheckstyleConfigurationsParser.java:125)
        at com.github.checkstyle.parser.CheckstyleConfigurationsParser.parse(CheckstyleConfigurationsParser.java:102)
        at com.github.checkstyle.Main.main(Main.java:186)
Caught: groovy.lang.GroovyRuntimeException: Error: !
groovy.lang.GroovyRuntimeException: Error: !
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at diff.executeCmd(diff.groovy:680)
        at diff.executeCmd(diff.groovy:673)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at diff$_generateDiffReport_closure5.doCall(diff.groovy:421)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at diff.generateDiffReport(diff.groovy:404)
        at diff$generateDiffReport$20.callCurrent(Unknown Source)
        at diff.run(diff.groovy:40)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

```

**Comment by rnveach** [comment]
@Kevin222004 Please make sure you are rebased on the lastest checkstyle master and using the latest contribution. XML exception is a known issue fixed a few days ago and you need to update the DTD on your regression config.

**Comment by Kevin222004** [comment]
DIff regression report:- https://kevin222004.github.io/RegressionReport/

**Comment by Kevin222004** [comment]
Please Review

**Comment by nrmancuso** [comment]
@Kevin222004 

> Please Review

Please make CI happy, and explain why we changed commits in orekit/hipparchus

**Comment by Kevin222004** [comment]
@nrmancuso the reason to change is it is giving the same failure as shown in https://github.com/checkstyle/checkstyle/pull/11673#issuecomment-1140458939 which is expected as our new changes 

previous contributor has sent the pr to orekit and it is merged https://github.com/CS-SI/Orekit/pull/30 but our commit has not been updated in validation.sh as his pr is not merged and if i talk about the Hipparchus,

according to https://github.com/checkstyle/checkstyle/blob/5497291e5cda75376570bef2832114258fa501f9/.ci/validation.sh#L625
its look to me that orekit is dependent on Hipparchus so i have changed both of the commit which are the recently merged.

**Comment by Kevin222004** [comment]
by mistake closed

**Comment by nrmancuso** [comment]
@Kevin222004 thanks for the explanation. Please send orekit/Hipparchus commit update as separate PR with `supplemental: update …` as commit message. Make sure to explain update in PR description.

**Comment by Kevin222004** [comment]
> @Kevin222004 thanks for the explanation. Please send orekit/Hipparchus commit update as separate PR with `supplemental: update …` as commit message. Make sure to explain update in PR description.

after changing the commit with latest one only failure is at https://github.com/CS-SI/Orekit/blob/4c193cffa7358ae2004db3219da244718384830f/src/main/java/org/orekit/gnss/RinexObservationLoader.java#L304-L306
line 304 UnnecessaryParentheses: Unnecessary parentheses around identifier 'parseString'. is expected as our new changes i will make pr on it by today

**Comment by Kevin222004** [comment]
I am not sure why this is failing https://dev.azure.com/romanivanovjr/romanivanovjr/_build/results?buildId=12704&view=logs&jobId=6a833393-c551-5000-0916-5000fd3b86e7&j=6a833393-c551-5000-0916-5000fd3b86e7&t=bdbbd129-77f0-5d73-ce7d-db58d34c8358

**Comment by nrmancuso** [comment]
@Kevin222004 note how old commits are separated by quotes, this is to trick spell check

**Comment by Kevin222004** [comment]
@nrmancuso ci is happy now :)

**Comment by Kevin222004** [comment]
@nrmancuso ci is happy now :)

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+         LAMBDA, TEXT_BLOCK_LITERAL_BEGIN, LAND, LITERAL_INSTANCEOF, GT, LT, GE, \
+         LE, EQUAL, NOT_EQUAL, UNARY_MINUS, UNARY_PLUS, INC, DEC, LNOT, BNOT, \
+         POST_INC, POST_DEC
+
+
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessaryparentheses;
+
+public class InputUnnecessaryParenthesesIdentifier {
```
Comment (nrmancuso) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java:16]: For each example please add  '// ok` version too:

```
    public static boolean isBad(Class<?> clazz) {
        return true
            && (test(clazz)); // violation 'Unnecessary parentheses around identifier 'test'.'
    }

    public static boolean isGood(Class<?> clazz) {
        return true
            && test(clazz); // ok
    }


```

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 2]
```diff
@@ -552,10 +552,16 @@ else if (isSurrounded(ast) && unnecessaryParenAroundOperators(ast)) {
      *         parentheses.
      */
     private static boolean isSurrounded(DetailAST ast) {
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:554]: Please update javadoc to be more simple; we do not need to know implementation details of a method. We really only need the first sentence:

```
Tests if the given {@code DetailAST} is surrounded by parentheses.

```

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Please extend test cases to include:
https://kevin222004.github.io/RegressionReport/spoon/xref/home/kevin/Desktop/Testing-Project-for-checkstyle-regression/spoon/spoon-master/spoon-control-flow/src/test/resources/control-flow/ControlFlowArithmetic.java.html#L428 (method call in `if` condition)
https://kevin222004.github.io/RegressionReport/spoon/xref/home/kevin/Desktop/Testing-Project-for-checkstyle-regression/spoon/spoon-master/src/test/resources/noclasspath/FileIO.java.html#L1263 (concatenation of method call return value)
https://kevin222004.github.io/RegressionReport/lombok-ast/xref/home/kevin/Desktop/Testing-Project-for-checkstyle-regression/lombok.ast/test/resources/performance/extra166y/AbstractParallelAnyArray.java.html#L5344 (typecast)


Items:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 3]
```diff
@@ -552,10 +552,16 @@ else if (isSurrounded(ast) && unnecessaryParenAroundOperators(ast)) {
...
+        // next sibling can't be other than right parentheses
+        // if the parent's tokentype is METHOD_CALL,
+        // check whether the parent's previous sibling is left parentheses.
         final DetailAST prev = ast.getPreviousSibling();
-        return prev != null && prev.getType() == TokenTypes.LPAREN;
+        final DetailAST parent = ast.getParent();
+        return prev != null && prev.getType() == TokenTypes.LPAREN
+                || parent.getType() == TokenTypes.METHOD_CALL
+                    && parent.getPreviousSibling() != null
+                        && parent.getPreviousSibling().getType() == TokenTypes.LPAREN;
```
[the remark this answers is not in the document]
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: @nrmancuso as of know i have done exact changes as you have described if you look it is good let me know else i will try to make it more simple

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java (hunk 1 above)
Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java:19]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 2 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:551]: done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java [hunk 4]
```diff
@@ -542,20 +542,21 @@ else if (isSurrounded(ast) && unnecessaryParenAroundOperators(ast)) {
...
      */
     private static boolean isSurrounded(DetailAST ast) {
-        // if previous sibling is left parenthesis,
-        // next sibling can't be other than right parenthesis
         final DetailAST prev = ast.getPreviousSibling();
-        return prev != null && prev.getType() == TokenTypes.LPAREN;
+        final DetailAST parent = ast.getParent();
+        final boolean isPreviousSiblingLeftParenthesis = prev != null
+                && prev.getType() == TokenTypes.LPAREN;
+        final boolean isChildOfMethodCallWith = parent.getType() == TokenTypes.METHOD_CALL
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:556]: @Kevin222004 please do not leave us hanging... method call with what? :) `isChildOfMethodCallWith...` was just an example, how about `isChildOfMethodCallWithSiblingParenthesis`? or maybe even better is just `isMethodCallWithUnnecessaryParenthesis`?

**Review by nrmancuso — CHANGES_REQUESTED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java (hunk 4 above)
Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java:556]: yes sorry my mistake I think `isMethodCallWithUnnecessaryParenthesis`

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: done

Comment (Kevin222004) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheck.java]: https://github.com/checkstyle/checkstyle/pull/12579#discussion_r1063619977 is it good

**Review by nrmancuso — APPROVED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        return true;
+    }
+
+    public void testIF() {
+        int a = 9;
+        int b = 10;
+        if (a > b) {
+            System.out.println("a is greater then b");
+        }
+        else if ((b < a) && ((8>9) || (!m3()))){ // 3 violations
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java:74]: Unless there is a reason, please split this into multiple lines to show the violation messages.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java (hunk 5 above)
Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java:74]: @rnveach  should i split all the code in which their are multiple violation

Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java]: this change has been 
done

Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java]: If you can do it without harming the tests, then yes.

Comment (Kevin222004) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java]: done

Comment (rnveach) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesIdentifier.java]: Thanks

**Review by rnveach — APPROVED** [review]

**Comment by rnveach** [comment]
@Kevin222004 None of the links work in https://kevin222004.github.io/RegressionReport/local-checkstyle/index.html

**Comment by Kevin222004** [comment]
> @Kevin222004 None of the links work in https://kevin222004.github.io/RegressionReport/local-checkstyle/index.html Did you miss uploading them?

I have directly copy paste the whole report but only one the link of local checkstyle is not working because i think the resource for that is  generated on my local setup of checkstyle and its generated in target folder with that path accordingly in HTML  or XML file. 

should i regenerate the regression with github action i have to see how i can share all the links for local checkstyle

**Comment by rnveach** [comment]
I wouldn't bother re-generating it. I looked over the other projects and I was fine with them.

**Pull request merged**